### ISSUE TITLE
Feature request: draw_string() on image types

### ISSUE THREAD

**Issue by pfeatherstone** [davisking/dlib#2264]
It would be great if there was a version of ```draw_string()``` that worked on image types like ```dlib::matrix<dlib::rgb_pixel>``` for example.
Might submit a PR at some point. This issue could serve as a reminder.

**Comment by arrufat** [comment]
Yes, I agree! I've also needed that feature some times! (to be honest I have to use OpenCV just for that...)

I've thought about it many times, but never got into it. Using the builtin [bdf](http://sofia.nmsu.edu/~mleisher/Software/otf2bdf/) font support would be a way to do it, but the main issue is that you need a font for each size you want (unless you don't care about proper scaling).

**Comment by pfeatherstone** [comment]
Looking at the code in ```gui_widget.cpp``` and ```font.cpp```, i see no reason why bits of code couldn't be ripped out and refactored for image types rather than canvas. Unless i've missed something...

**Comment by arrufat** [comment]
Yes, as I said, I never looked into it... Maybe it's straightforward :)

**Comment by pfeatherstone** [comment]
Might give it a quick go now.

**Comment by pfeatherstone** [comment]
[main.cpp.txt](https://github.com/davisking/dlib/files/5710327/main.cpp.txt)

**Comment by pfeatherstone** [comment]
That worked for me. It's a basis. Probs need to template to support different image types, pixel types and probably other stuff

**Comment by pfeatherstone** [comment]
Also support different fonts. Here i just use the default font

**Comment by pfeatherstone** [comment]
Yeah so this is the function that works for me:
```
void draw_string(matrix<rgb_pixel>& img, dlib::point start_point, std::string str, dlib::rgb_pixel color)
{
    const dlib::rectangle rect(start_point,start_point);
    
    auto f_ptr = default_font::get_font();   
    const font& f = *f_ptr;
    
    long y_offset = rect.top() + f.ascender() - 1;

    long pos = rect.left()+f.left_overflow();
    for (string::size_type i = 0; i < str.size(); ++i)
    {
        // ignore the '\r' character
        if (str[i] == '\r')
            continue;
        // A combining character should be applied to the previous character, and we
        // therefore make one step back. If a combining comes right after a newline, 
        // then there must be some kind of error in the string, and we don't combine.
        if(is_combining_char(str[i]) && 
           pos > rect.left() + static_cast<long>(f.left_overflow()))
        {
            pos -= f[str[i]].width();
        }

        if (str[i] == '\n')
        {
            y_offset += f.height();
            pos = rect.left()+f.left_overflow();
            continue;
        }

        // only look at letters in the intersection area
        if (img.nr() + static_cast<long>(f.height()) < y_offset)
        {
            // the string is now below our rectangle so we are done
            break;
        }
        else if (0 > pos - static_cast<long>(f.left_overflow()) && 
            pos + static_cast<long>(f[str[i]].width() + f.right_overflow()) < 0 )
        {
            pos += f[str[i]].width();                
            continue;
        }
        else if (img.nc() + static_cast<long>(f.right_overflow()) < pos)
        {
            // keep looking because there might be a '\n' in the string that
            // will wrap us around and put us back into our rectangle.
            continue;
        }

        // at this point in the loop we know that f[str[i]] overlaps 
        // horizontally with the intersection rectangle area.

        const letter& l = f[str[i]];
        for (unsigned short j = 0; j < l.num_of_points(); ++j)
        {
            const long x = l[j].x + pos;
            const long y = l[j].y + y_offset;
            // draw each pixel of the letter if it is inside the intersection
            // rectangle
            if (x >= 0 && x < img.nc() && y >= 0 && y < img.nr())
                assign_pixel(img(y,x), color);
        }

        pos += l.width();
    }
}
```

**Comment by pfeatherstone** [comment]
By the way, it's almost a carbon copy of what's in fonts.cpp

**Comment by arrufat** [comment]
Cool! I've tried with another font and it also works, but it doesn't support UTF-8 (I wrote hello in Korean: 안녕하세요).

That's one of the reasons I use this [font](https://github.com/Dalgona/neodgm-pro) instead, when building UI apps using dlib, using [`dlib::ustring`](http://dlib.net/parsing.html#ustring))
![draw_string](https://user-images.githubusercontent.com/1671644/102512182-133a6700-40cd-11eb-9624-9fc671ac38e7.png)

Thanks for starting working on this :)

**Comment by davisking** [comment]
Ha, yeah, I’ve periodically wanted this too but never enough to bother with it :)

**Comment by arrufat** [comment]
Actually, just changing the function signature to use `dlib::ustring` instead of `std::string` and passing a `dlib::convert_utf8_to_utf32(str)` worked. This makes me really happy!

So maybe we could add an overload of `dlib::ustring` like `dlib::label` has for setting the text.
![testfont](https://user-images.githubusercontent.com/1671644/102570001-d3ee3380-4129-11eb-8b2b-42141c04e3cb.png)

**Comment by pfeatherstone** [comment]
Just use something like this (exactly what's done in fonts.h):
```
template <typename T, typename traits, typename alloc, typename pixel_type>
void draw_string (
    ...,
    const std::basic_string<T,traits,alloc>& str,
    const pixel_type& color,
    ...
)
```

**Comment by pfeatherstone** [comment]
@arrufat i've still got the FFT PR on the go. Have you fleshed this out enough to submit a PR? If not, don't worry, I'll submit a PR at a later time.

**Comment by arrufat** [comment]
@pfeatherstone Oh, I will give it a go, but I can't promise anything. It will be a good chance to better know the GUI API. Thank you! :)

**Comment by pfeatherstone** [comment]
Errh, you can't use ```draw_string()``` without GUI support. In fact, I think you can't do any image transformations without GUI support now. I think that's a bug.

**Comment by davisking** [comment]
Eh, yeah, forgot about that bit of optionality :(

**Comment by davisking** [comment]
Anyway, this should be fixed.  An easy fix is to just put this code in some other header file.  The `font` class could also be moved to a separate header that doesn't depend on gui code.  There isn't any reason why that class needs to depend on any of that stuff.  Either of you guys want to do this?

**Comment by pfeatherstone** [comment]
I can do but i would rather wait for the FFT stuff to go through first. Then i will start branching my fork so i can have a couple of PRs on the go in the future (if necessary)

**Comment by arrufat** [comment]
I'll try to do it soon, then. Let's see if I can find some time before @pfeatherstone finishes his other PR.

**Comment by arrufat** [comment]
Maybe I am missing something, but it doesn't seem that trivial to me. The font class has methods defined that depend on `canvas`, which depends on the `gui_core` library. It looks like a lot of refactoring, right?

**Comment by davisking** [comment]
Just `draw_string()` right?  It doesn't seem like there is any reason that function can't just be a global exactly like the one you just added rather than being a member function.

**Comment by arrufat** [comment]
Maybe it's too late here, I'll give it a go again later. Thank you :)

**Comment by davisking** [comment]
Yeah, it's an API change.  Probably few users actually call that.  A better alternative is to make the canvas a template argument so the concrete `canvas` type doesn't need to be `#included`

**Comment by davisking** [comment]
That way the APi says exactly the same from the user point of view but the header that defines the canvas doesn't need to be included.

**Comment by arrufat** [comment]
I guess this can now be closed. It has been addressed in PRs #2270 and #2273.
Thank you @pfeatherstone!
